test/interface/test_models.py

```diff
@@ -116,3 +119,126 @@ def test_multiple_models_ambiquous_spec(self, session):
+        query = session.query(Foo).join(Bar)
+        assert get_default_model(query) is None
+
+    def test_empty_query(self, session):
+        query = session.query()
+        assert get_default_model(query) is None
+
+
+class TestAutoJoin:
+
+    def test_model_not_present(self, session, db_uri):
+        query = session.query(Foo)
+        query = auto_join(query, 'Bar')
+
+        join_type = "INNER JOIN" if "mysql" in db_uri else "JOIN"
+
+        expected = (
+            "SELECT "
+            "foo.id AS foo_id, foo.name AS foo_name, "
+            "foo.count AS foo_count, foo.bar_id AS foo_bar_id \n"
+            "FROM foo {join} bar ON bar.id = foo.bar_id".format(join=join_type)
+        )
+        assert str(query) == expected
+
+    def test_model_already_present(self, session):
+        query = session.query(Foo, Bar)
+
+        # no join applied
+        expected = (
+            "SELECT "
+            "foo.id AS foo_id, foo.name AS foo_name, "
+            "foo.count AS foo_count, foo.bar_id AS foo_bar_id, "
+            "bar.id AS bar_id, bar.name AS bar_name, bar.count AS bar_count \n"
+            "FROM foo, bar"
+        )
+        assert str(query) == expected
+
+        query = auto_join(query, 'Bar')
+        assert str(query) == expected   # no change
+
+    def test_model_already_joined(self, session, db_uri):
+        query = session.query(Foo).join(Bar)
+
+        join_type = "INNER JOIN" if "mysql" in db_uri else "JOIN"
+
+        expected = (
+            "SELECT "
+            "foo.id AS foo_id, foo.name AS foo_name, "
+            "foo.count AS foo_count, foo.bar_id AS foo_bar_id \n"
+            "FROM foo {join} bar ON bar.id = foo.bar_id".format(join=join_type)
+        )
+        assert str(query) == expected
+
+        query = auto_join(query, 'Bar')
+        assert str(query) == expected   # no change
+
+    def test_model_eager_joined(self, session, db_uri):
+        query = session.query(Foo).options(joinedload(Foo.bar))
+
+        join_type = "INNER JOIN" if "mysql" in db_uri else "JOIN"
+
+        expected_eager = (
+            "SELECT "
+            "foo.id AS foo_id, foo.name AS foo_name, "
+            "foo.count AS foo_count, foo.bar_id AS foo_bar_id, "
+            "bar_1.id AS bar_1_id, bar_1.name AS bar_1_name, "
+            "bar_1.count AS bar_1_count \n"
+            "FROM foo LEFT OUTER JOIN bar AS bar_1 ON bar_1.id = foo.bar_id"
+        )
+        assert str(query) == expected_eager
+
+        expected_joined = (
+            "SELECT "
+            "foo.id AS foo_id, foo.name AS foo_name, "
+            "foo.count AS foo_count, foo.bar_id AS foo_bar_id, "
+            "bar_1.id AS bar_1_id, bar_1.name AS bar_1_name, "
+            "bar_1.count AS bar_1_count \n"
+            "FROM foo {join} bar ON bar.id = foo.bar_id "
+            "LEFT OUTER JOIN bar AS bar_1 ON bar_1.id = foo.bar_id".format(
+                join=join_type
+            )
+        )
+
+        # effect of filter?
```

Comment:
is this a todo of some sort?

Reply:
It is a TODOne ;)

Thanks, will remove